docs: add ceph telemetry design doc (#10254)
rook/rook: design/ceph/ceph-telemetry.md

  classes are added in the future.
  - `rook-num-rbd-storage-classes`
  - `rook-num-cephfs-storage-classes`
  - `rook-num-nfs-storage-classes`
  - `rook-num-bucket-storage-classes`
- `rook-num-storage-class-device-sets` - Number of storage class device sets.
- `rook-num-portable-storage-class-device-sets` - .
- `rook-mons-are-pvc` - true/false whether mons are on PVC.
- `rook-network-provider` - The network provider used for the cluster. e.g., default, host, multus.
- `rook-external-mode` - true/false if the cluster is in external mode.

## Information Rook is interested in that is already collected by Ceph telemetry
- RBD pools (name is stripped), some config info, and the number of them.
- Ceph Filesystems and MDS info.
- RGW count, zonegroups, zones.
- RBD mirroring info.

## Metrics that Rook would like to see added to Ceph's telemetry
- Number of NFS clusters
  - Ceph can count NFS clusters by listing the namespaces in the `.nfs` pool and report those
    namespaces that have `conf-nfs.<namespace>` objects present
- Filesystem mirroring info
- RGW bucket notification usage (Yuval may have good detail on this point)

## Proposed integration strategy

> - Rook will first merge [`rook-version`](https://github.com/rook/rook/pull/10161).

Reviewer: Since it's already merged, shall we generalize this section for future merges?

> - Rook, with input from Ceph telemetry, will approve a version of this design doc and treat all

Reviewer: Instead of locking on this design doc, I'm thinking the telemetry const names could be the source of coordination. Design docs are just never perfectly updated with the code.

Author: In most cases, I would agree. But since this is a contract we will maintain with the Ceph team, I think we should treat this doc as gospel in this case. Once Ceph implements these, we should never change them.

Reviewer: That would mean we could only add metrics to this design doc if they are implemented in a given batch, right? But capturing future possible metrics seems like a good and necessary thing. 

How strict of a contract is really necessary? If the values are just all key-value pairs, and the telemetry will only gather the keys we've told them about, it seems sufficient to look at the consts in the code. Even if we remove metrics in the future, I expect it won't break the telemetry, it would just not be gathered anymore.

  noted metric names as requirements.
- Ceph will add all of the noted telemetry keys from the config-key database to its telemetry and

>   backport to Ceph Pacific and Ceph Quincy.

Reviewer: ```suggestion
  backport to supported Ceph major versions.
```

  - Ceph code should handle the case where Rook has not set these fields.
  - Ceph code can add `rook-version` immediately if they expect it might get into the next
    Pacific/Quincy release more quickly.
- Rook will implement each metric (or related metric group) individually over time.

> This strategy will allow Rook time to add telemetry items as it is able without rushing. Because the

Reviewer: I wonder if a Ceph tracker will be the best way to signal that we are ready for them to pick up the next batch of rook updates, this would be good to confirm.

telemetry fields will be approved all at once, it will also minimize the coordination that is
required between Ceph and Rook. The Ceph team will not need to create PRs one-to-one with Rook, and
we can limit version mismatch issues as the telemetry is being added.

> ## Rejected metrics

Reviewer: PR comments could keep track of rejected metrics. Why track them in the design doc?

Author: While that is true, PR comments are often hidden, especially for PRs with lots of comments. This way, the distillation can be seen by anyone looking at the doc without having to find the PR and then un-hide all the comments from who knows how long ago. In my experience, a good design doc must mention discarded ideas so that people are less likely to suggest and later implement ideas that should be rejected once we all have forgotten what was rejected and why.

### Count of each CR Kind
Count of each type of CR: cluster, object, file, object store, mirror, bucket topic,
bucket notification, etc.

This was rejected for version one for a few reasons:
1. In most cases, the Ceph telemetry should be able to report this implicitly via its information.
  - Block pools, filesystems, and object stores can already be inferred easily
1. It would require a config-key for each CR sub-resource, and Rook has many.
1. This could be simpler if we had a json blob for the value, but the Ceph telemetry team has set a
   guideline not to do this.

We can revisit this on a case-by-case basis for specific CRs or features. For example, we may wish
to have ideas about COSI usage when that is available.

### Pod memory/CPU requests/limits
The memory/CPU requests/limits set on Ceph daemon types.

> This was rejected for a few reasons:

Reviewer: I don't see it as rejected or accepted yet, I'd suggest it still needs more discussion. I don't think the ceph telemetry prohibits us from collecting since it's really a small list in reality where the same resource limits are generally applied for all OSDs. Maybe we even ignore the resources set on storageClassDeviceSets and just collect the general `spec.resources`. This data doesn't have to be exhaustive, it's just to give us an idea of how features are being used and perhaps give us input to how we might change the defaults or implement features in the future.

Author: Laura specifically told us on the meeting yesterday not to put data structures into the config-key database. We need to clear anything like this with the telemetry team.

Reviewer: I do think this would be good to consider, so perhaps it's just how to format the metrics, even if we have a number of individual metrics for memory/cpu requests/limits for mons, osds, mgr, etc.

Reviewer 2: Can you please share an example of exactly how such metrics look like?
If we format the metrics in a way that it does not grow dynamically, and it is not too big, it should be fine.

Reviewer: We discussed offline that this may not be very useful to gather, so we will hold off on it for now anyway.

# Adding Rook data to Ceph telemetry
The solution plan agreed upon with the telemetry team is for the Rook operator to add telemetry to
the Ceph mon `config-key` database, and Ceph will read each of those items for telemetry retrieval.

## Guidelines from the Ceph telemetry team
- Users must opt in to telemetry
- Users must also opt into the new (delta) telemetry items added between Ceph versions
- Rook should not add information to `config-key` keys that can grow arbitrarily large to keep space
  usage of the mon database low

## Proposed metrics for Rook to collect
We propose naming metrics to indicate a hierarchy that can be parsed to add it to the Ceph telemetry
collection in a more ordered fashion.

For example `rook/kubernetes/version` would should be put into a JSON structure like shown:
```json
"rook": {
  "kubernetes": {
    "version"
  }
}
```

### Overall metrics
- `rook-version` - Rook version.

>   - **DISCUSS:** should we add `rook/version` or rename this to maintain a consistent hierarchy?

Reviewer: Do the ceph metrics all follow this pattern using paths? If all our metrics start with `rook/`, seems like we should rename to `rook/version`.

Reviewer 3: `rook/version` sounds better with metrics starts with `rook`

>   - **DISCUSS:** should we add `rook/version` or rename this to maintain a consistent hierarchy?

Reviewer 4: The name `rook/version` sounds right to me. IMO, the end report should have a single `rook` key that users/devs can search for, under which they can find all of the relevant information. The rook version number is a higher-level piece of information, so it could look something like this (the report gets alphabetized):

```json
"rook": {
  "csi": {
     "version"
   },
  "kubernetes": {
    "node-count",
    "version"
   },
   "version"
}
```

Reviewer 2: +1 for `rook/version`

> - `rook/kubernetes/version` - Kubernetes version.

Reviewer: The Kubernetes version is sufficient from Rook's perspective.

  - Ceph already collects os/kernel versions.
- `rook/kubernetes/node-count` - Number of Kubernetes nodes.
- `rook/csi/version` - Ceph CSI version.
- Number of CSI plugin pods running - This gives idea of how many k8s nodes are in the k8s cluster that
  are allowed to use Ceph storage.

>   - It is easiest count plugin pods separately since each DaemonSet is a different label. Plus, it

Reviewer 4: ```suggestion
  - It is easiest to count plugin pods separately since each DaemonSet is a different label. Plus, it
```

    can show if some CSI drivers are ever disabled.
  - `rook/csi/rbd-plugin-pod-count`
  - `rook/csi/cephfs-plugin-pod-count`
  - `rook/csi/nfs-plugin-pod-count`

### CephCluster metrics

> - `rook/cluster/nodes-with-ceph-pods-count` - Number of nodes on which the cluster has running pods.

Reviewer: Another metric to consider:
- `rook/cluster/mon/max-id` - The highest mon id, which increases as mons failover

  - Since clusters with portable PVCs have one "node" per PVC, this will help show the actual node
    count for Rook installs in portable environments

>   - We can get this by counting the number of crash collector pods; however, if any users disable

Reviewer 4: ```suggestion
  - We can get this by counting the number of crash collector pods; however, if any users disables
```

Author: I believe "... users disable..." is still correct. Or do you propose this should be "...user disables..."?

    the crash collector, we should report an unknown value, possibly `unknown`.
- `rook/cluster/storage-class/count` - Number of storage classes related to the current Ceph cluster. We can
  also break this up by specific kinds of storage classes, but the overall may be useful if new
  classes are added in the future.
  - `rook/cluster/storage-class/rbd-count`
  - `rook/cluster/storage-class/cephfs-count`
  - `rook/cluster/storage-class/nfs-count`

>   - `rook/cluster/storage-class/bucket-count`

Reviewer 4: Is this the total count? If so, how about `rook/cluster/storage-class/total-count`?

Or you could even break this up into:
`rook/cluster/storage-class/count/total`
`rook/cluster/storage-class/count/rbd`
`rook/cluster/storage-class/count/cephfs`
...etc

That way, all the "counts" could be easily grouped together in case any other key gets added under `storage-class`.

> - `rook/cluster/storage-class-device-set/count` - Total number of storage class device sets.

Reviewer: for brevity maybe we don't need the full name from the CR. "device set" is a common short name.
```suggestion
- `rook/cluster/device-set/count` - Total number of storage class device sets.
```

- `rook/cluster/storage-class-device-set/portable-count` - Number of storage class device sets that
  are configured to be portable.

> - `rook/cluster/storage-class-device-set/non-portable-count` - Number of storage class device sets

Reviewer 4: Same for this. You could say `rook/cluster/storage-class-device-set/total-count` (if this is indeed the total), or

`rook/cluster/storage-class-device-set/count/total`
`rook/cluster/storage-class-device-set/count/portable`
`rook/cluster/storage-class-device-set/count/non-portable`

  that are not configured to be portable.

> - `rook/cluster/mons/are-pvc` - true/false whether mons are on PVC.

Reviewer: ```suggestion
- `rook/cluster/mons/use-pvc` - true/false whether mons are on PVC.
```

  - Leave some ability to expand the mon section in the future if desired.
- `rook/cluster/network-provider` - The network provider used for the cluster. e.g., default, host, multus.
- `rook/cluster/external-mode` - true/false if the cluster is in external mode.

## Information Rook is interested in that is already collected by Ceph telemetry
- RBD pools (name is stripped), some config info, and the number of them.
- Ceph Filesystems and MDS info.
- RGW count, zonegroups, zones.
- RBD mirroring info.

## Metrics that Rook would like to see added to Ceph's telemetry
- Number of NFS clusters
  - Ceph can count NFS clusters by listing the namespaces in the `.nfs` pool and report those
    namespaces that have `conf-nfs.<namespace>` objects present
- Filesystem mirroring info

> - RGW bucket notification usage (Yuval may have good detail on this point)

Reviewer 4: Any specific data points would be helpful to know here (understood that this is a WIP design concept though)

Reviewer: @BlaineEXE How about if we track follow-up additions as issues? I'm afraid it will get lost in this doc. Issues at least remind us occasionally to do something from the stale bot.

- `rook/kubernetes/...`
  - `rook/kubernetes/version` - Kubernetes version.
    - Ceph already collects os/kernel versions.
  - `rook/kubernetes/node/count/total` - Number of Kubernetes nodes.
- `rook/csi/...`
  - `rook/csi/version` - Ceph CSI version.
  - `rook/csi/plugin-pod/...` - Info about CSI plugin pods
    - `rook/csi/plugin-pod/count/...` - Info about how many CSI plugin pods exist and of what types
      This gives idea of how many k8s nodes are in the k8s cluster that are allowed to use Ceph
      storage. With different types, telemetry can show if some CSI drivers are ever disabled.
      - `rook/csi/plugin-pod/count/rbd`
      - `rook/csi/plugin-pod/count/cephfs`
      - `rook/csi/plugin-pod/count/nfs`

### CephCluster metrics
- `rook/cluster/node/count/with-ceph-pods` - Number of nodes on which the cluster has running pods.
  - Since clusters with portable PVCs have one "node" per PVC, this will help show the actual node
    count for Rook installs in portable environments
  - We can get this by counting the number of crash collector pods; however, if any users disable
    the crash collector, we will report `-1` to represent "unknown"

> - `rook/cluster/storage-class/...` - Info about storage classes related to the current Ceph cluster

Reviewer: Since everything is part of the same cluster, how about if we remove "cluster" from the path? This simplifies the path for most of the metrics.
```suggestion
- `rook/storage-class/...` - Info about storage classes related to the current Ceph cluster
```

# Adding Rook data to Ceph telemetry
The solution plan agreed upon with the telemetry team is for the Rook operator to add telemetry to
the Ceph mon `config-key` database, and Ceph will read each of those items for telemetry retrieval.

## Guidelines from the Ceph telemetry team
- Users must opt in to telemetry
- Users must also opt into the new (delta) telemetry items added between Ceph versions
- Rook should not add information to `config-key` keys that can grow arbitrarily large to keep space
  usage of the mon database low (limited growth is still acceptable)

## Metrics for Rook to collect

> Metric names will indicate a hierarchy that can be parsed to add it to  Ceph telemetry collection in

Reviewer: ```suggestion
Metric names will indicate a hierarchy that can be parsed to add it to Ceph telemetry collection in
```

a more ordered fashion.

> For example `rook/version` and `rook/kubernetes/version` should be put into a structure like shown:

Reviewer: ```suggestion
For example `rook/version` and `rook/kubernetes/version` would be put into a structure like shown:
```

```json
"rook": {
  "version": "vx.y.z"
  "kubernetes": {
    "version": "vX.Y.Z"
  }
}
```

### Overall metrics
- `rook/version` - Rook version.
- `rook/kubernetes/...`
  - `rook/kubernetes/version` - Kubernetes version.

>     - Ceph already collects os/kernel versions.

Reviewer: I'm curious if Ceph actually collects these in a K8s environment as expected. Or maybe you said you already verified that.

Author: You can check out the sample I got from a minikube test cluster: https://gist.github.com/BlaineEXE/2fafaaeec065803390ebd75b498c1a59

I think it largely only gathers the in-container information and not information about kubernetes' bare hosts. The must-gather has CentOS Stream 8 as the OS, which definitely isn't what minikube uses, but is I think what the Ceph images use.

Currently, I don't know if there would be a good way to collect the actual raw host info, but it would seem useful if we can. I would leave this for a follow-up unless someone has good ideas for collecting the info.

  - `rook/kubernetes/node/count/total` - Number of Kubernetes nodes.
- `rook/csi/...`
  - `rook/csi/version` - Ceph CSI version.
  - `rook/csi/plugin-pod/...` - Info about CSI plugin pods
    - `rook/csi/plugin-pod/count/...` - Info about how many CSI plugin pods exist and of what types
      This gives idea of how many k8s nodes are in the k8s cluster that are allowed to use Ceph
      storage. With different types, telemetry can show if some CSI drivers are ever disabled.
      - `rook/csi/plugin-pod/count/max` - Max number of pods of any CSI plugin type

>       - `rook/csi/plugin-pod/count/types` - Number of CSI plugin types. This can helpfully show if a

Reviewer: How is this helpful if we already have the rook version?

Author: I think it could be. I think it at least saves us from having to look up or have memorized what version of Rook added new features. I think that seems beneficial.

Reviewer: Thinking more since our discussion, I still don't see the benefit. The report itself will show if rbd/ceph/nfs are being used or not. That seems useful enough by itself. Gathering the count of `types` falls in the category of info that tells us there is a disconnect between what rook reports and what ceph gathers, but doesn't help us answer any questions about how rook is being used. If telemetry doesn't tell us how rook is actively being used, I would vote not to collect it.

        newer version of Rook supports a new plugin pod type than the Ceph telemetry version is
        aware of.
      - `rook/csi/plugin-pod/count/rbd`
      - `rook/csi/plugin-pod/count/cephfs`
      - `rook/csi/plugin-pod/count/nfs`

### CephCluster metrics

> - `rook/cluster/node/count/with-ceph-pods` - Number of nodes on which the cluster has running pods.

Reviewer: Shall we merge with the k8s node metrics? 
```suggestion
- `rook/kubernetes/node/count/with-ceph-pods` - Number of nodes on which the cluster has running pods.
```

Author: I think that still seems like a cluster metric to me. It will be different from cluster-to-cluster, so I think that makes it a cluster metric.

Reviewer: Everything is a cluster metrics, right? If the operator is managing different clusters, all the metrics are still reported to this instance of ceph, and each instance of ceph would phone home separately to report telemetry. So if we're going to have metrics about nodes, seems like they should all be under the same tree, whether it's `rook/cluster` or `rook/kubernetes` it probably doesn't matter.

Author: Sure, but for CSI, it seems straightforward to just count the plugin pods under the `csi` heading. It's really only the "nodes with ceph pods" idea that is a little bit abstract/funny.

Author: I also disagree that all node information should be under the same heading. Inherently, the total number of Kubernetes nodes is platform information, which I would think should be under a different heading than information about the Rook-Ceph cluster itself.

But I suppose it depends on whose "view" is most important. Since this is Ceph telemetry, I think Ceph's view is the makes the most sense. In that view, I see platform information separately.

But if we take the Kubernetes view, then everything is in comparison the the Kubernetes cluster, and what you say would make sense. But I don't think that "view" is the right one for Ceph telemetry.

    - `rook/cluster/storage-class/count/types` - Number of storage class types. This can helpfully
      show if a newer version of Rook supports a new plugin pod type than the Ceph telemetry version
      is aware of
    - `rook/cluster/storage-class/count/total` - This is additionally useful in the case of a
      newly-added storage class type not recognized by an older Ceph telemetry version
    - `rook/cluster/storage-class/count/rbd`
    - `rook/cluster/storage-class/count/cephfs`
    - `rook/cluster/storage-class/count/nfs`
    - `rook/cluster/storage-class/count/bucket`
- `rook/cluster/storage/...` - Info about storage configuration
  - `rook/cluster/storage/device-set/...` - Info about storage class device sets
    - `rook/cluster/storage/device-set/count/...` - Number of device sets of given types
      - `rook/cluster/storage/device-set/count/total`
      - `rook/cluster/storage/device-set/count/portable`
      - `rook/cluster/storage/device-set/count/non-portable`
- `rook/cluster/mons/...` - Info about monitors and mon configuration
  - `rook/cluster/mons/use-pvc` - true/false whether mons are on PVC

>   - `rook/cluster/mons/max-id` - The highest mon ID, which increases as mons fail over

Reviewer: How about adding mon count?

Reviewer: A bool for stretch cluster would be helpful, unless ceph already indicates that.

Reviewer: Oh right, the existing ceph info should be sufficient for the mons.

  - `rook-num-csi-nfs-plugin-pods`
- `rook-num-nodes` - Number of nodes on which the cluster has running pods.
  - Since clusters with portable PVCs have one "node" per PVC, this will help show the actual node
    count for Rook installs in portable environments
  - We can get this by counting the number of crash collector pods; however, if any users disable
    the crash collector, we should report an unknown value, possibly `unknown`.
- `rook-num-storage-classes` - Number of storage classes related to the current Ceph cluster. We can
  also break this up by specific kinds of storage classes, but the overall may be useful if new
  classes are added in the future.
  - `rook-num-rbd-storage-classes`
  - `rook-num-cephfs-storage-classes`
  - `rook-num-nfs-storage-classes`
  - `rook-num-bucket-storage-classes`
- `rook-num-storage-class-device-sets` - Number of storage class device sets.
- `rook-num-portable-storage-class-device-sets` - .
- `rook-mons-are-pvc` - true/false whether mons are on PVC.
- `rook-network-provider` - The network provider used for the cluster. e.g., default, host, multus.
- `rook-external-mode` - true/false if the cluster is in external mode.

> **DISCUSS:** Travis suggested: Count of each type of CR: cluster, object, file, object store,

Reviewer 5: Maybe we can also think of  osd data,

How much consumption and memory is utilized,
It's already there in the dashboard but this can be good to add with the telemetry too

Author: Example telemetry: https://gist.github.com/BlaineEXE/2fafaaeec065803390ebd75b498c1a59

Reviewer 6: memory usage already exists, but would it be useful to record the kubernetes resource constraints for memory and CPU?

Author: @richardelling I had considered this, but I opted not to include it. I'm not sure how useful that information really is. We can certainly add it, but I can't think of any good insights that would give us that are more helpful to the Ceph or Rook projects than the size of the usage information already collected by Ceph telemetry. It's also possible the telemetry team would consider this information to be sensitive similarly to how they consider the ceph config values sensitive. 

It's also a hard metric to collect since that can change for OSDs for each device set, and we would need several keys. Even just given the complexity, I think we should only push to add this if we can provide rationale for its use.

Reviewer 6: agree on all points. Speaking for ourselves, we already track and manage it through our existing telemetry service.

Reviewer: @richardelling agreed, I'd vote for capturing all of the `resources` from the cluster CR and other applicable CRs.

Reviewer: For the original question, as discussed offline we don't necessarily need the counts of all the CRs since most of that can be implied by other metrics (ceph metrics already collected for number of pools, object stores, etc)

# Adding Rook data to Ceph telemetry
The solution plan agreed upon with the telemetry team is for the Rook operator to add telemetry to
the Ceph mon `config-key` database, and Ceph will read each of those items for telemetry retrieval.
See: https://github.com/rook/rook/pull/10161

## Guidelines from the Ceph telemetry team
- Users must opt in to telemetry
- Users must also opt into the new (delta) telemetry items added between Ceph versions
- Rook should not add information to `config-key` keys that can grow arbitrarily large to keep space
  usage of the mon database low.

## Proposed metrics for Rook to collect
- `rook-version` - Rook version.

>   - Done in https://github.com/rook/rook/pull/10161

Reviewer: Instead of referring to PRs when each one is implemented, how about if we put all the telemetry name consts in a single place (perhaps create a telemetry.go), and link it from this doc?

- `rook-kubernetes-version` - Kubernetes version.

>   - This is similar to the already collected os/kernel versions.

Reviewer: What do you mean it's similar to the os version? Maybe the k8s version is just self-explanatory enough and I'm missing the point.

Author: Ceph telemetry already collects the OS info. I would consider the k8s version at the same level in terms of how identifying the information is, so I guess to me it seems relevant.

> - `rook-num-kubernetes-nodes` - Number of Kubernetes nodes.

Reviewer: ```suggestion
- `rook-kubernetes-node-count` - Number of Kubernetes nodes.
```

> - `rook-ceph-csi-version` - Ceph CSI version.

Reviewer: ceph seems implicit, and this is also more consistent with later csi metric names that don't include "ceph" in the name
```suggestion
- `rook-csi-version` - Ceph CSI version.
```

- Num CSI plugin pods running - This gives idea of how many k8s nodes are in the k8s cluster that
  are allowed to use Ceph storage.
  - It is easiest count plugin pods separately since each DaemonSet is a different label. Plus, it
    can show if some CSI drivers are ever disabled.

>   - `rook-num-csi-rbd-plugin-pods`

Reviewer: how about a `count` suffix instead of "num" for all the applicable metrics?
```suggestion
  - `rook-csi-rbd-plugin-count`
```

  - `rook-num-csi-cephfs-plugin-pods`
  - `rook-num-csi-nfs-plugin-pods`

> - `rook-num-nodes` - Number of nodes on which the cluster has running pods.

Reviewer: ```suggestion
- `rook-nodes-with-ceph-pods-count` - Number of nodes on which the cluster has running pods.
```

  - Since clusters with portable PVCs have one "node" per PVC, this will help show the actual node
    count for Rook installs in portable environments
  - We can get this by counting the number of crash collector pods; however, if any users disable
    the crash collector, we should report an unknown value, possibly `unknown`.

> - `rook-num-storage-classes` - Number of storage classes related to the current Ceph cluster. We can

Reviewer: Why do we need the aggregate number if we collect the individual counts?
```suggestion
- Number of storage classes related to the current Ceph cluster. We can
```

Author: I think we want to keep the overall count. This could be important if we add a new storage type in the future that isn't identified by a specific type.

  also break this up by specific kinds of storage classes, but the overall may be useful if new
  classes are added in the future.

>   - `rook-num-rbd-storage-classes`

Reviewer: ```suggestion
  - `rook-rbd-storage-class-count`
```

  - `rook-num-cephfs-storage-classes`
  - `rook-num-nfs-storage-classes`
  - `rook-num-bucket-storage-classes`

> - `rook-num-storage-class-device-sets` - Number of storage class device sets.

Reviewer: Was the intent for this one to be the total count? How about if we just make it the non-portable count?
```suggestion
- `rook-non-portable-storage-class-device-set-count` - Number of storage class device sets.
```

Author: Why get rid of the overall count and only have portable/non-portable variants? We can compute the non-portable type by comparing portable to overall. And I think the overall is better at-a-glance info. Plus it more closely follows how Ceph already collects its telemetry. Note that "count" is a total, with further breakdown after.
```
    "mon": {
        "count": 1,
        # ...
        "ipv4_addr_mons": 2,
        "ipv6_addr_mons": 0,
        "min_mon_release": 17,
        "v1_addr_mons": 1,
        "v2_addr_mons": 1
    },
 ```
 
 We can also collect all 3, but following Ceph's convention, I am reluctant to not include the overall.

> - `rook-num-portable-storage-class-device-sets` - .

Reviewer: ```suggestion
- `rook-portable-storage-class-device-set-count` - .
```

- `rook-mons-are-pvc` - true/false whether mons are on PVC.
- `rook-network-provider` - The network provider used for the cluster. e.g., default, host, multus.
- `rook-external-mode` - true/false if the cluster is in external mode.

> 

Reviewer: Perhaps we could have sections of metrics in this doc so we can see what's related to the cluster CR, or overall metrics, etc? Then it may be easier to see if metrics are needed for particular CRs, and so on.

Author: Should we add all consts at once, or only when implemented?

Reviewer: I would expect the consts would only be added when implemented.
